### DISCUSSION TITLE
Schema runtimeWiring too slow

### DISCUSSION THREAD

**Discussion by gobimcp** [graphql-java/graphql-java#2791] [General]
We build dynamic schemas based on the tables in the db and the schemas are relatively complex. We also do `RuntimeWiring.newRuntimeWiring()` and its taking around 1 sec on average now, which is too slow for our API. Any recommendations to improve the wiring time? If there is a way to serialize the wired schema, that would be great so we can cache it to make our APIs very fast. 

Thanks for your suggestions.

**Comment by jord1e** [comment]
You recreate the wiring on every request?

**Comment by gobimcp** [comment]
@jord1e  Yes. Because we have dynamic schema for each account, we will need to wire it for each request. Do you have any suggestions not to avoid doing it for every request?

**Comment by jord1e** [comment]
Have you taken a look at Field Visibilty?

https://www.graphql-java.com/documentation/v17/field-visibility

https://github.com/graphql-java/graphql-java/tree/master/src/main/java/graphql/schema/visibility

You basically generate the wiring and typedefinitionregistry etc one time and then selectively determine your available field set per operation

You can create a new GraphQL instance by transforming your "everything instance"

```java
masterInstanceContainingEverything.transform(builder -> builder.schema(graphQL.getGraphQLSchema()
        .transform(schemaBuilder ->
                schemaBuilder.codeRegistry(graphQL.getGraphQLSchema().getCodeRegistry()
                        .transform(codeRegistryBuilder ->
                                codeRegistryBuilder.fieldVisibility(new NoIntrospectionGraphqlFieldVisibility())))))); // Insert your own per-user visibility
```

Also caching the GraphQL instances and load balancing your requests to the same server based on your user

**Comment by gobimcp** [comment]
@jord1e Thanks for the suggestion here. We will have close to a million accounts and each has different fields in the schema. At a first look, looks like creating `everything instance` and selectively determining the available fields per account id may result in much bigger master object to handle and setting visibility also may take a while because almost all the fields will be hidden except for around 15 or so object types per account. Anyway, let me think little bit more on whether we can use this field visibility in any other way to make this work.

**Comment by jord1e** [comment]
I assume you have introspection enabled on a per-client basis?

If not you could always just write an instrumentation that blocks the fields specific users shouldn't access.

Maybe a maintainer has better ideas :) - a million different accounts (so schemas) is above my knowledge

**Comment by gobimcp** [comment]
I thought about it little bit and I think we will run several issues such as duplicates. For example, several accounts can have an object type named `products` but each can have different set of fields inside that object type, so maintaining a master instance is hard because of such duplicates.

**Comment by gobimcp** [comment]
@bbakerman Do you have any suggestions here?

**Reply by bbakerman** [reply]
A graphql schema is meant to be a complete type system and set of code for fetching data.  As such its nominally expensive to build BUT that cost is meant to be apportioned across many requests.

Why does it take so long you ask - well it has lot of rules that need to be enforced (a directive used on an element MUST be in valid locations say - to not do those checks means that the type system integrity is compromised) and just generally it takes a fair bit of time to traverse a large tree structure that is the "type system".

So the idea that you have 1 million schemas that are different is not where graphql comes from.  As a weak analogy, do you build one million copies of your application code or do you provision one million different sql databases per customer?  because a graphql type system is a complex description of data and has a cost.

I think the best ideas would be


* amortize these costs
** cache in memory the built schema per customer
*** surely not every customer is on every time?
** be prepared for lots of memory to be used
** dont make a new runtime fresh - take an existing schema and change it a little bit if you can
*** have some sort of master schema - then "transform" it per customer to the small set of differences
* ** This still costs time - you need to traverse all the schema elements even if you delete only some so I am not sure it will pay off

There is no way to serialise a schema?  Why?  Well the deserialization costs are up there in the same order of magnitude as just printing it to SDL text and re-parsing it.

BTW what version of graphql-java are you on - we made significant improvements in schema build time in v16 - albeit large schemas take a fair bit of time (eg not millieconds but 100s of milliseconds)

**Reply by gchomatas** [reply]
Thank you @bbakerman for the great suggestions!

> As a weak analogy, do you build one million copies of your application code or do you provision one million different sql databases per customer?

The beauty of GraphQL is its schema and the great editing experience editors like GraphiQL can provide through introspection. We could provide a generalized schema that would look like function calls or calling the REST API through GraphQL fields, e.g. `object(type: "book", properties: ["title", "description", ...])`instead of `book { title, description, holder {name, ...}`. This would somehow diminish its value and usability. Maintaining a user friendly custom schema per customer makes it super easy to use their data by just inspecting the schema or use autocompletion. Of course there is a high performance penalty to pay but what I am trying to say is that schema multi tenancy is a valid use case. A generalized customer facing schema is not always the best solution as it could be when using a single database for all customers in the backend (which is what we do anyway for maintaining all custom object types :) )

**Comment by gobimcp** [comment]
Thanks @bbakerman for your thoughts. We already use V17.x so we have seen really huge improvements in the wiring time as part of that upgrade (right now our median wiring time is ~450 ms and 99th percentile is ~1 sec whereas before it used be around 3 secs at 99th percentile).

We already cache the wired schema in memory and the problem is we need more memory as you mentioned to cache all our schemas.

**Comment by gchomatas** [comment]
All of the schemas we serve share a considerable set of "standard object types". Users may add more properties to their standard types (relatively few in comparison to how many already exist in the standard type by default) or they can introduce completely new types (most users introduce just a few and with considerably less fields than the standard objects).
So an efficient solution could be to keep a prewired master schema with just the standard object types shared across all customers and then generate schema instances by somehow cloning the prewired schema and wiring the extra properties and types that are specific to each customer.

@bbakerman if I understood well from the comments above, the idea was to keep a mega schema and switch off fields. Is there a way to do the opposite, i.e. start with a prewired schema (with the common object types) that we clone and add the missing fields? 

BTW we are performing the wiring from pre-generated and cached SDLs and we also cache the wired schemas but the wired schemas are quite big and we cannot fit a lot in memory or keep them in a distributed cache since they are not serializable.

**Reply by jbellenger** [reply]
> Is there a way to do the opposite, i.e. start with a prewired schema (with the common object types) that we clone and add the missing fields?

Have you seen that you  can initialize a new GraphQLSchema from an existing one, like
```
        GraphQLSchema
          .newSchema(standardSchema)
          .additionalTypes(extraTypes)
          .build()
```

I think you could combine this with your wiring to create a new GraphQL object.

**Reply by gchomatas** [reply]
Thanks @jbellenger! Adding the new types is straightforward but we will need to traverse the standard object types in the standard schema too to clone them and add the extra properties users may have added. We could try that. I am not sure how much performance gain we will eventually get but it might worth to try it.

**Comment by bbakerman** [comment]
>  the idea was to keep a mega schema and switch off fields. Is there a way to do the opposite, i.e. start with a prewired schema (with the common object types) that we clone and add the missing fields?

Yeah we have a mechanism to "edit" an existing schema allowing you to add / delete / change elements.

It is nominally faster than building it from scratch however depending on how many edits you do, the time maybe in the same order of magnitude.  You would need to try it and see.

`graphql.schema.SchemaTransformer` uses a visitor pattern and the visitor gives back "edit instructions" on what to do to a schema.

An example bit of code that uses it is `graphql.schema.transform.FieldVisibilitySchemaTransformation` which allows you to remove fields (and linked types) from a schema.

Also `graphql.schema.SchemaTransformerTest` has more examples and so does https://www.graphql-java.com/documentation/schema

You challenge will be how "standard" you master template schema is - if its 90% the same maybe you will be in a good place - if its not then more work will need to be done and that equals time.

**Comment by Stefanie899** [comment]
I've managed to get this working to some extent, but have an additional question:

**Is it possible to go from a printed schema (as a string) to a static (unwired) schema?** 

As far as I can tell, it doesn't seem possible but I wanted to make sure. 

Obviously we could just wire up the custom part of the schema, but this seems to be considerably slower, as I have to both wire the custom part of the schema, and then traverse that schema and transform any reference to types that already exist in the built-in schema so there aren't conflicts (as well as having to introduce some new interfaces between the built-in and custom wired schemas to avoid conflicts, as replacing interfaces through the transformer doesn't seem to work). This is all in addition to inserting the new types/fields in the built-in schema and attaching the appropriate data fetchers.

Alternatively, I suppose we could use the `TypeRegistry` from parsing the printed schema and potentially build the appropriate GraphQLTypes from there, but that seems like it will be a pain.

**Reply by bbakerman** [reply]
> Is it possible to go from a printed schema (as a string) to a static (unwired) schema?

Its eminently possible

```
GraphQLSchema graphQLSchema = SchemaGenerator.createdMockedSchema("Your SDL");
```

This will  create a schema that has no runtime wiring associated with it - well it does - its got a mock one that does nothing

```
graphql.schema.idl.RuntimeWiring#MOCKED_WIRING
```

**Reply by bbakerman** [reply]
> Obviously we could just wire up the custom part of the schema, but this seems to be considerably slower, as I have to both wire the custom part of the schema, and then traverse that schema and transform any reference to types that already exist in the built-in schema so there aren't conflicts (as well as having to introduce some new interfaces between the built-in and custom wired schemas to avoid conflicts, as replacing interfaces through the transformer doesn't seem to work). 

The fact is if you HEAVILY edit the schema structure then you will take about as much time as it does to created one from scratch.

See above for all the reasons on why this is
